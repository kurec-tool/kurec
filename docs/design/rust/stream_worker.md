# Rust ストリームワーカー実装ルール

## 基本方針

ストリームワーカーは、イベント駆動型アーキテクチャにおいて「何らかの入力（SSEまたはJetStreamストリーム）を読んでイベントメッセージを処理し、次のストリームにイベントメッセージを送る」というパターンを実装するためのコンポーネントです。以下のルールに従って実装してください。

## 実装ルール

### 1. 責務の分離

- **イベントメッセージの読み取り・書き込み処理**はインフラ層の責務
  - `EventSource` と `EventSink` インターフェースを使用
  - 具体的な実装は `JsSubscriber`、`JsPublisher`、`MirakcSseSource` など
- **イベントに対する処理**はドメイン層の責務
  - `StreamHandler` インターフェースを実装
  - ドメインロジックを含むハンドラクラスを作成

### 2. ハンドラの実装

- ハンドラは `StreamHandler` トレイトを実装する
- 戻り値は `Result<Option<OutputEvent>, Error>` の形式
  - `Ok(Some(output_event))`: 処理成功、出力イベントあり
  - `Ok(None)`: 処理成功、出力イベントなし
  - `Err(error)`: 処理失敗
- 依存関係は `Arc<dyn Interface>` の形でコンストラクタに渡す

```rust
#[async_trait]
impl StreamHandler<InputEvent, OutputEvent, MyError> for MyHandler {
    async fn handle(&self, event: InputEvent) -> Result<Option<OutputEvent>, MyError> {
        // イベント処理ロジック
    }
}
```

### 3. エラーハンドリング

- エラー型は `ClassifyError` トレイトを実装する
- `error_action()` メソッドで `ErrorAction::Retry` または `ErrorAction::Ignore` を返す
- リトライ可能なエラー（インフラ層のエラーなど）は `ErrorAction::Retry` を返す
- リトライ不可能なエラー（バリデーションエラーなど）は `ErrorAction::Ignore` を返す
- JetStreamはAckを返さなければ自動的にリトライするため、この機構を活用する

```rust
impl ClassifyError for MyError {
    fn error_action(&self) -> ErrorAction {
        match self {
            MyError::Repository(_) => ErrorAction::Retry,
            MyError::Validation(_) => ErrorAction::Ignore,
        }
    }
}
```

### 4. ワーカーの構築

- `StreamWorker::new()` を使用してワーカーを構築する
- `source`、`sink`、`handler` を引数に渡す
- 必要に応じて `durable()` または `durable_auto()` でdurable名を設定する
- 必要に応じて `with_middleware()` でミドルウェアを追加する

```rust
let worker = StreamWorker::new(source, sink, handler)
    .durable("my-worker")
    .with_middleware(LoggingMiddleware::new());
```

### 5. シャットダウン処理

- `CancellationToken` を使用してグレースフルシャットダウンを実装する
- `run()` メソッドに `shutdown_token` を渡す
- シャットダウン時は処理中のイベントを完了させてから終了する

```rust
worker.run(shutdown_token).await?;
```

### 6. 簡易実装（マクロ使用）

- 単純なケースでは `#[stream_worker]` マクロを使用できる
- マクロは関数に適用し、`process_event_worker` という名前の関数を生成する

```rust
#[stream_worker]
async fn process_event(event: InputEvent) -> Result<OutputEvent, MyError> {
    // イベント処理ロジック
}
```

## テスト戦略

### 1. ハンドラのテスト

- ハンドラの依存関係をモックして単体テストを実施
- `StreamHandler::handle()` メソッドの戻り値を検証

```rust
#[tokio::test]
async fn test_handler() {
    // モックの作成
    let mock_repository = MockRepository::new();
    let mock_notifier = MockNotifier::new();
    
    // ハンドラの作成
    let handler = MyHandler::new(
        Arc::new(mock_repository),
        Arc::new(mock_notifier),
    );
    
    // テスト対象のイベント
    let event = InputEvent { /* ... */ };
    
    // ハンドラの実行
    let result = handler.handle(event).await;
    
    // 結果の検証
    assert!(result.is_ok());
    if let Ok(Some(output)) = result {
        assert_eq!(output.field, expected_value);
    }
}
```

### 2. ワーカーのテスト

- `JsSubscriber` と `JsPublisher` のモックを作成
- `StreamWorker` を構築して実行
- 出力イベントを検証

```rust
#[tokio::test]
async fn test_worker() {
    // モックの作成
    let mock_source = MockEventSource::new();
    let mock_sink = MockEventSink::new();
    let handler = Arc::new(MyHandler::new(/* ... */));
    
    // ワーカーの構築
    let worker = StreamWorker::new(
        Arc::new(mock_source),
        Arc::new(mock_sink),
        handler,
    );
    
    // シャットダウントークンの作成
    let shutdown = CancellationToken::new();
    let shutdown_clone = shutdown.clone();
    
    // ワーカーを別スレッドで実行
    let handle = tokio::spawn(async move {
        worker.run(shutdown_clone).await.unwrap();
    });
    
    // テスト対象のイベントを送信
    mock_source.send_event(InputEvent { /* ... */ }).await;
    
    // 少し待機
    tokio::time::sleep(Duration::from_millis(100)).await;
    
    // シャットダウン
    shutdown.cancel();
    handle.await.unwrap();
    
    // 結果の検証
    let events = mock_sink.get_published_events();
    assert_eq!(events.len(), 1);
    assert_eq!(events[0].field, expected_value);
}
```

## 注意点

1. **メモリリーク**:
   - `Arc` を使ったサイクリック参照に注意する
   - 長時間実行されるワーカーではメモリ使用量をモニタリングする

2. **パフォーマンス**:
   - 重い処理はブロッキングしないようにする
   - 必要に応じてバッチ処理を検討する

3. **エラー伝播**:
   - エラーは適切に伝播させ、ログに記録する
   - エラーの原因を特定しやすくするために、コンテキスト情報を付加する

4. **冪等性**:
   - イベント処理は冪等であるべき（同じイベントを複数回処理しても結果が変わらない）
   - 特にリトライ可能なエラーの場合、同じイベントが複数回処理される可能性がある
