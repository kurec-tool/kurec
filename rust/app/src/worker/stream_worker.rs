use anyhow::Result;
use async_trait::async_trait;
use domain::event::Event;
use domain::ports::{event_sink::EventSink, event_source::EventSource};
use futures::future::BoxFuture;
use futures::StreamExt;
use shared_core::error_handling::{ClassifyError, ErrorAction};
use std::marker::PhantomData;
use std::sync::Arc;
use tokio::select;
use tokio_util::sync::CancellationToken;
use tracing::{error, warn};

/// ストリームワーカーのミドルウェアトレイト
/// 入力イベントを処理して出力イベントを生成する前後に処理を挟むことができる
#[async_trait]
pub trait StreamMiddleware<I, O, E>: Send + Sync + 'static
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    // 戻り値を Option<O> に変更
    async fn handle(&self, event: I, next: StreamNext<'_, I, O, E>) -> Result<Option<O>, E>;
}

/// ミドルウェアチェーンの次の処理を表す構造体
pub struct StreamNext<'a, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    // ハンドラの型と戻り値を Option<O> に変更
    pub(crate) handler:
        Arc<dyn Fn(I) -> BoxFuture<'static, Result<Option<O>, E>> + Send + Sync + 'static>,
    _phantom: PhantomData<&'a ()>,
}

impl<I, O, E> StreamNext<'_, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    // ハンドラの型と戻り値を Option<O> に変更
    pub fn new(
        handler: Arc<dyn Fn(I) -> BoxFuture<'static, Result<Option<O>, E>> + Send + Sync + 'static>,
    ) -> Self {
        Self {
            handler,
            _phantom: PhantomData,
        }
    }

    // 戻り値を Option<O> に変更
    pub async fn run(&self, event: I) -> Result<Option<O>, E> {
        (self.handler)(event).await
    }
}

impl<I, O, E> Clone for StreamNext<'_, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    // ハンドラの型を Option<O> に変更
    fn clone(&self) -> Self {
        Self {
            handler: self.handler.clone(),
            _phantom: PhantomData,
        }
    }
}

/// イベントハンドラトレイト
#[async_trait]
pub trait StreamHandler<I, O, E>: Send + Sync + 'static
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    // 戻り値を Option<O> に変更
    async fn handle(&self, event: I) -> Result<Option<O>, E>;
}

/// 関数をハンドラとして扱うためのラッパー
pub struct FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    // 関数の戻り値を Option<O> に変更
    F: Fn(I) -> BoxFuture<'static, Result<Option<O>, E>> + Send + Sync + 'static,
{
    f: F,
    _phantom: PhantomData<(I, O, E)>,
}

impl<I, O, E, F> FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    // 関数の戻り値を Option<O> に変更
    F: Fn(I) -> BoxFuture<'static, Result<Option<O>, E>> + Send + Sync + 'static,
{
    pub fn new(f: F) -> Self {
        Self {
            f,
            _phantom: PhantomData,
        }
    }
}

#[async_trait]
impl<I, O, E, F> StreamHandler<I, O, E> for FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    // 関数の戻り値を Option<O> に変更
    F: Fn(I) -> BoxFuture<'static, Result<Option<O>, E>> + Send + Sync + 'static,
{
    // 戻り値を Option<O> に変更
    async fn handle(&self, event: I) -> Result<Option<O>, E> {
        (self.f)(event).await
    }
}

/// ストリームワーカー
/// 入力イベントを処理して出力イベントを生成するワーカー
pub struct StreamWorker<I, O, IErr, E>
where
    I: Event,
    O: Event,
    IErr: Send + Sync + 'static,
    E: ClassifyError + Send + Sync + 'static,
{
    source: Arc<dyn EventSource<I, IErr>>,
    sink: Arc<dyn EventSink<O>>,
    handler: Arc<dyn StreamHandler<I, O, E>>,
    middlewares: Vec<Arc<dyn StreamMiddleware<I, O, E>>>,
    durable_name: Option<String>,
}

impl<I, O, IErr, E> StreamWorker<I, O, IErr, E>
where
    I: Event,
    O: Event,
    IErr: Send + Sync + 'static,
    E: ClassifyError + Send + Sync + 'static,
{
    /// 新しいStreamWorkerを作成
    pub fn new(
        source: Arc<dyn EventSource<I, IErr>>,
        sink: Arc<dyn EventSink<O>>,
        handler: Arc<dyn StreamHandler<I, O, E>>,
    ) -> Self {
        Self {
            source,
            sink,
            handler, // Arc<dyn StreamHandler> を受け取る
            middlewares: Vec::new(),
            durable_name: None,
        }
    }

    /// ミドルウェアを追加
    pub fn with_middleware<M>(mut self, middleware: M) -> Self
    where
        // M の戻り値を Option<O> に変更
        M: StreamMiddleware<I, O, E> + 'static,
    {
        self.middlewares.push(Arc::new(middleware));
        self
    }

    /// durable名を設定
    pub fn durable(mut self, name: &str) -> Self {
        self.durable_name = Some(name.to_string());
        self
    }

    /// durable名を自動生成
    pub fn durable_auto(mut self) -> Self {
        let type_name = std::any::type_name::<I>();
        let last_segment = type_name.split("::").last().unwrap_or(type_name);
        self.durable_name = Some(format!("worker_{}", last_segment));
        self
    }

    /// ミドルウェアチェーンを構築して実行
    // シグネチャと戻り値を変更
    async fn execute_middleware_chain(
        handler: Arc<dyn StreamHandler<I, O, E>>, // F -> Arc<dyn StreamHandler>
        middlewares: &[Arc<dyn StreamMiddleware<I, O, E>>],
        event: I,
    ) -> Result<Option<O>, E> {
        // Result<O, E> -> Result<Option<O>, E>
        // ミドルウェアがない場合は直接ハンドラを実行
        if middlewares.is_empty() {
            return handler.handle(event).await; // handler(event) -> handler.handle(event)
        }

        // ミドルウェアチェーンを構築
        let mut chain = Vec::new();
        for middleware in middlewares {
            chain.push(Arc::clone(middleware));
        }

        // 最後のミドルウェアの次の処理はハンドラ
        let handler_clone = handler.clone(); // handler は Arc なので clone するだけ
                                             // ハンドラの型と戻り値を Option<O> に変更
        let handler_fn = Arc::new(move |e: I| -> BoxFuture<'static, Result<Option<O>, E>> {
            let handler_inner = handler_clone.clone();
            Box::pin(async move { handler_inner.handle(e).await }) // handler_inner(e) -> handler_inner.handle(e)
        });

        // ミドルウェアチェーンを逆順に実行
        let mut next = StreamNext::new(handler_fn);
        for middleware in chain.into_iter().rev() {
            let prev_next = next.clone();
            let middleware_clone = Arc::clone(&middleware);
            // ハンドラの型と戻り値を Option<O> に変更
            let next_handler = Arc::new(move |e: I| -> BoxFuture<'static, Result<Option<O>, E>> {
                let pn = prev_next.clone();
                let middleware_inner = Arc::clone(&middleware_clone);
                Box::pin(async move { middleware_inner.handle(e, pn).await })
            });
            next = StreamNext::new(next_handler);
        }

        // 最初のミドルウェアを実行
        next.run(event).await
    }

    /// ワーカーを実行
    pub async fn run(self, shutdown: CancellationToken) -> Result<()> {
        // sourceからメッセージストリームを取得
        let mut stream = self.source.subscribe().await?;
        let shutdown_token = shutdown.clone();

        // ハンドラとミドルウェアをArcでラップ
        let handler = self.handler;
        let middlewares: Vec<Arc<dyn StreamMiddleware<I, O, E>>> = self.middlewares;
        let sink = self.sink;

        // メッセージ処理ループ
        loop {
            select! {
                // シャットダウントークンが発火したら終了
                _ = shutdown_token.cancelled() => {
                    break;
                }
                // メッセージを受信したら処理
                message = stream.next() => {
                    match message {
                        Some(Ok(event)) => {
                            // 通常の処理
                            let result = Self::execute_middleware_chain(
                                handler.clone(),
                                &middlewares,
                                event
                            ).await;

                            match result {
                                Ok(Some(output_event)) => {
                                    // 出力イベントをSinkに発行
                                    if let Err(e) = sink.publish(output_event).await {
                                        error!(error = %e, "Failed to publish output event");
                                        // パブリッシュに失敗した場合は再試行
                                        continue;
                                    }
                                }
                                Ok(None) => {
                                    // 出力イベントがない場合は何もしない
                                }
                                Err(e) => {
                                    // エラーアクションに基づいて処理
                                    match e.error_action() {
                                        ErrorAction::Retry => {
                                            // 再試行の場合はログを出力
                                            warn!(error = %e, "Error processing event, will retry");
                                            continue;
                                        }
                                        ErrorAction::Ignore => {
                                            // エラーを無視する場合はログを出力
                                            warn!(error = %e, "Error processing event, ignoring");
                                        }
                                    }
                                }
                            }
                        }
                        Some(Err(error)) => {
                            // エラーの処理
                            // ログは既にソース側で記録済み

                            // 必要に応じて追加の処理
                            // metrics::increment_counter!("event_errors", "type" => error.type_str());
                        }
                        None => {
                            // ストリームが終了
                            break;
                        }
                    }
                }
            }
        }

        Ok(())
    }
}
