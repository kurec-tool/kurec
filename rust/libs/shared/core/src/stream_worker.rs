use crate::error_handling::ClassifyError;
use crate::event_metadata::Event;
use crate::event_publisher::EventPublisher;
use crate::event_subscriber::EventSubscriber;
use anyhow::Result;
use async_trait::async_trait;
use futures::future::BoxFuture;
use futures::StreamExt;
use std::marker::PhantomData;
use std::sync::Arc;
use tokio::select;
use tokio_util::sync::CancellationToken;

/// ストリームワーカーのミドルウェアトレイト
/// 入力イベントを処理して出力イベントを生成する前後に処理を挟むことができる
#[async_trait]
pub trait StreamMiddleware<I, O, E>: Send + Sync + 'static
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    async fn handle(&self, event: I, next: StreamNext<'_, I, O, E>) -> Result<O, E>;
}

/// ミドルウェアチェーンの次の処理を表す構造体
pub struct StreamNext<'a, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    pub(crate) handler: Arc<dyn Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static>,
    _phantom: PhantomData<&'a ()>,
}

impl<'a, I, O, E> StreamNext<'a, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    pub fn new(
        handler: Arc<dyn Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static>,
    ) -> Self {
        Self {
            handler,
            _phantom: PhantomData,
        }
    }

    pub async fn run(&self, event: I) -> Result<O, E> {
        (self.handler)(event).await
    }
}

impl<'a, I, O, E> Clone for StreamNext<'a, I, O, E>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    fn clone(&self) -> Self {
        Self {
            handler: self.handler.clone(),
            _phantom: PhantomData,
        }
    }
}

/// イベントハンドラトレイト
#[async_trait]
pub trait StreamHandler<I, O, E>: Send + Sync + 'static
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
{
    async fn handle(&self, event: I) -> Result<O, E>;
}

/// 関数をハンドラとして扱うためのラッパー
pub struct FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    F: Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static,
{
    f: F,
    _phantom: PhantomData<(I, O, E)>,
}

impl<I, O, E, F> FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    F: Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static,
{
    pub fn new(f: F) -> Self {
        Self {
            f,
            _phantom: PhantomData,
        }
    }
}

#[async_trait]
impl<I, O, E, F> StreamHandler<I, O, E> for FnStreamHandler<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    F: Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static,
{
    async fn handle(&self, event: I) -> Result<O, E> {
        (self.f)(event).await
    }
}

/// ストリームワーカー
/// 入力イベントを処理して出力イベントを生成するワーカー
pub struct StreamWorker<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    F: Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static + Clone,
{
    subscriber: Arc<dyn EventSubscriber<I>>,
    publisher: Arc<dyn EventPublisher<O>>,
    handler: F,
    middlewares: Vec<Arc<dyn StreamMiddleware<I, O, E>>>,
    durable_name: Option<String>,
}

impl<I, O, E, F> StreamWorker<I, O, E, F>
where
    I: Event,
    O: Event,
    E: ClassifyError + Send + Sync + 'static,
    F: Fn(I) -> BoxFuture<'static, Result<O, E>> + Send + Sync + 'static + Clone,
{
    /// 新しいStreamWorkerを作成
    pub fn new(
        subscriber: Arc<dyn EventSubscriber<I>>,
        publisher: Arc<dyn EventPublisher<O>>,
        handler: F,
    ) -> Self {
        Self {
            subscriber,
            publisher,
            handler,
            middlewares: Vec::new(),
            durable_name: None,
        }
    }

    /// ミドルウェアを追加
    pub fn with_middleware<M>(mut self, middleware: M) -> Self
    where
        M: StreamMiddleware<I, O, E> + 'static,
    {
        self.middlewares.push(Arc::new(middleware));
        self
    }

    /// durable名を設定
    pub fn durable(mut self, name: &str) -> Self {
        self.durable_name = Some(name.to_string());
        self
    }

    /// durable名を自動生成
    pub fn durable_auto(mut self) -> Self {
        let type_name = std::any::type_name::<I>();
        let last_segment = type_name.split("::").last().unwrap_or(type_name);
        self.durable_name = Some(format!("worker_{}", last_segment));
        self
    }

    /// ミドルウェアチェーンを構築して実行
    async fn execute_middleware_chain(
        handler: F,
        middlewares: &[Arc<dyn StreamMiddleware<I, O, E>>],
        event: I,
    ) -> Result<O, E> {
        // ミドルウェアがない場合は直接ハンドラを実行
        if middlewares.is_empty() {
            return handler(event).await;
        }

        // ミドルウェアチェーンを構築
        let mut chain = Vec::new();
        for middleware in middlewares {
            chain.push(Arc::clone(middleware));
        }

        // 最後のミドルウェアの次の処理はハンドラ
        let handler_clone = handler.clone();
        let handler_fn = Arc::new(move |e: I| -> BoxFuture<'static, Result<O, E>> {
            let handler_inner = handler_clone.clone();
            Box::pin(async move { handler_inner(e).await })
        });

        // ミドルウェアチェーンを逆順に実行
        let mut next = StreamNext::new(handler_fn);
        for middleware in chain.into_iter().rev() {
            let prev_next = next.clone();
            let middleware_clone = Arc::clone(&middleware);
            let next_handler = Arc::new(move |e: I| -> BoxFuture<'static, Result<O, E>> {
                let pn = prev_next.clone();
                let middleware_inner = Arc::clone(&middleware_clone);
                Box::pin(async move { middleware_inner.handle(e, pn).await })
            });
            next = StreamNext::new(next_handler);
        }

        // 最初のミドルウェアを実行
        next.run(event).await
    }

    /// ワーカーを実行
    pub async fn run(self, shutdown: CancellationToken) -> Result<()> {
        // サブスクライバーからメッセージストリームを取得
        let mut stream = self.subscriber.subscribe().await?;
        let shutdown_token = shutdown.clone();

        // ハンドラとミドルウェアをArcでラップ
        let handler = self.handler.clone();
        let middlewares: Vec<Arc<dyn StreamMiddleware<I, O, E>>> = self.middlewares;
        let publisher = self.publisher;

        // メッセージ処理ループ
        loop {
            select! {
                // シャットダウントークンが発火したら終了
                _ = shutdown_token.cancelled() => {
                    break;
                }
                // メッセージを受信したら処理
                message = stream.next() => {
                    match message {
                        Some((event, ack)) => {
                            // ミドルウェアチェーンを実行
                            let result = Self::execute_middleware_chain(
                                handler.clone(),
                                &middlewares,
                                event,
                            ).await;

                            match result {
                                Ok(output_event) => {
                                    // 出力イベントをパブリッシュ
                                    publisher.publish(output_event).await?;
                                    ack.ack().await?;
                                }
                                Err(e) => {
                                    // エラーアクションに基づいて処理
                                    match e.error_action() {
                                        crate::error_handling::ErrorAction::Retry => {
                                            // nack（再試行）
                                            // JetStreamの場合、ackしないと自動的に再配信される
                                        }
                                        crate::error_handling::ErrorAction::Ignore => {
                                            // エラーを無視してack
                                            ack.ack().await?;
                                        }
                                    }
                                }
                            }
                        }
                        None => {
                            // ストリームが終了したら終了
                            break;
                        }
                    }
                }
            }
        }

        Ok(())
    }
}
