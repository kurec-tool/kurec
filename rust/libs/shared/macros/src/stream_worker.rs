use proc_macro::TokenStream;
use proc_macro2::Span;
use quote::quote;
use syn::{
    parse_macro_input, punctuated::Punctuated, Expr, FnArg, ItemFn, Lit, Meta, MetaNameValue,
    PatType, ReturnType, Token, Type,
};

/// 関数の入力パラメータ型を取得する
fn extract_input_type(args: &Punctuated<FnArg, Token![,]>) -> Option<&Type> {
    if args.len() != 1 {
        return None;
    }

    match &args[0] {
        FnArg::Typed(PatType { ty, .. }) => Some(ty),
        _ => None,
    }
}

/// 関数の戻り値型からResultの中身を取得する
fn extract_output_and_error_types(return_type: &ReturnType) -> Option<(&Type, &Type)> {
    match return_type {
        ReturnType::Type(_, ty) => {
            // Result<OutputType, ErrorType>の形式を期待
            if let Type::Path(type_path) = ty.as_ref() {
                let path = &type_path.path;
                if path.segments.len() == 1 && path.segments[0].ident == "Result" {
                    if let syn::PathArguments::AngleBracketed(args) = &path.segments[0].arguments {
                        if args.args.len() == 2 {
                            let output_type = match &args.args[0] {
                                syn::GenericArgument::Type(t) => t,
                                _ => return None,
                            };
                            let error_type = match &args.args[1] {
                                syn::GenericArgument::Type(t) => t,
                                _ => return None,
                            };
                            return Some((output_type, error_type));
                        }
                    }
                }
            }
            None
        }
        _ => None,
    }
}

/// #[stream_worker]マクロの実装
pub fn stream_worker_impl(attr: TokenStream, item: TokenStream) -> TokenStream {
    // 関数を解析
    let input_fn = parse_macro_input!(item as ItemFn);
    let fn_name = &input_fn.sig.ident;
    let fn_vis = &input_fn.vis;

    // 属性パラメータを解析
    let args = parse_macro_input!(attr with Punctuated<Meta, Token![,]>::parse_terminated);

    let mut durable_name = None;

    for meta in args {
        if let Meta::NameValue(MetaNameValue { path, value, .. }) = meta {
            if path.is_ident("durable") {
                if let Expr::Lit(expr_lit) = &value {
                    if let Lit::Str(lit_str) = &expr_lit.lit {
                        durable_name = Some(lit_str.value());
                    }
                }
            }
        }
    }

    // 関数の入力型と出力型を取得
    let input_type = match extract_input_type(&input_fn.sig.inputs) {
        Some(ty) => ty,
        None => {
            return syn::Error::new(
                Span::call_site(),
                "Function must have exactly one parameter",
            )
            .to_compile_error()
            .into();
        }
    };

    let (output_type, error_type) = match extract_output_and_error_types(&input_fn.sig.output) {
        Some((output, error)) => (output, error),
        None => {
            return syn::Error::new(
                Span::call_site(),
                "Function must return Result<OutputType, ErrorType>",
            )
            .to_compile_error()
            .into();
        }
    };

    // durableメソッドの呼び出しを生成
    let durable_method = if let Some(name) = durable_name {
        let name_lit = syn::LitStr::new(&name, Span::call_site());
        quote! { .durable(#name_lit) }
    } else {
        quote! { .durable_auto() }
    };

    // StreamWorkerを構築して実行するコードを生成
    let expanded = quote! {
        // 元の関数をそのまま保持
        #input_fn

        // ワーカーを実行する関数を生成
        #fn_vis async fn #fn_name _worker(
            js_ctx: &shared_infra::jetstream::JetStreamCtx,
            shutdown: tokio_util::sync::CancellationToken
        ) -> anyhow::Result<()> {
            use std::sync::Arc;
            use shared_core::event_metadata::HasStreamDef;
            use shared_core::stream_worker::StreamWorker;
            use shared_infra::jetstream::{JsPublisher, JsSubscriber};

            // サブスクライバーとパブリッシャーを作成
            let subscriber = Arc::new(JsSubscriber::<#input_type>::new(
                js_ctx.clone(),
                <#input_type as HasStreamDef>::stream_name(),
                <#input_type as HasStreamDef>::stream_subject(),
            ));

            let publisher = Arc::new(JsPublisher::<#output_type>::new(
                js_ctx.clone(),
                <#output_type as HasStreamDef>::stream_name(),
                <#output_type as HasStreamDef>::stream_subject(),
            ));

            // ハンドラ関数をラップ
            let handler = |event: #input_type| -> futures::future::BoxFuture<'static, std::result::Result<#output_type, #error_type>> {
                Box::pin(async move {
                    #fn_name(event).await
                })
            };

            // StreamWorkerを構築して実行
            StreamWorker::new(subscriber, publisher, handler)
                #durable_method
                .run(shutdown)
                .await
                .map_err(|e| anyhow::anyhow!("Worker error: {}", e))
        }
    };

    TokenStream::from(expanded)
}
