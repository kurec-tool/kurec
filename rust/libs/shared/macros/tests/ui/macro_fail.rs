use shared_macros::stream_worker;

// 入力パラメータがない関数（エラーになるべき）
#[stream_worker]
async fn invalid_function_no_params() -> Result<(), anyhow::Error> {
    Ok(())
}

// 複数の入力パラメータを持つ関数（エラーになるべき）
#[stream_worker]
async fn invalid_function_multiple_params(
    param1: String,
    param2: String,
) -> Result<(), anyhow::Error> {
    Ok(())
}

// Result型を返さない関数（エラーになるべき）
#[stream_worker]
async fn invalid_function_not_result(param: String) -> String {
    param
}

fn main() {}
