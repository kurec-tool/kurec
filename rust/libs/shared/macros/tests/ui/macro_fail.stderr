error: custom attribute panicked
 --> tests/ui/macro_fail.rs:3:1
  |
3 | #[event(subject = "foo.bar")] // ← stream が無いのでコンパイル失敗
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: stream parameter is required

error[E0601]: `main` function not found in crate `$CRATE`
 --> tests/ui/macro_fail.rs:4:16
  |
4 | pub struct Bad;
  |                ^ consider adding a `main` function to `$DIR/tests/ui/macro_fail.rs`
